go/vt/vtgate/executor.go

```diff
@@ -1461,6 +1465,31 @@ func (e *Executor) getPlan(vcursor *vcursorImpl, sql string, comments sqlparser.
 	return e.checkThatPlanIsValid(stmt, plan)
 }
 
+func (e *Executor) canNormalizeStatement(stmt sqlparser.Statement, qo iQueryOption, setVarComment string) bool {
+	return (e.normalize && sqlparser.CanNormalize(stmt)) ||
+		sqlparser.MustRewriteAST(stmt, qo.getSelectLimit() > 0) || setVarComment != ""
+}
+
+func prepareSetVarComment(vcursor *vcursorImpl, stmt sqlparser.Statement) (string, error) {
+	if vcursor == nil || vcursor.Session().InReservedConn() {
+		return "", nil
+	}
+
+	if !vcursor.Session().HasSystemVariables() {
+		return "", nil
+	}
+	if _, supportsOptimizerHint := stmt.(sqlparser.SupportOptimizerHint); !supportsOptimizerHint {
+		vcursor.NeedsReservedConn()
+		return "", nil
+	}
+
+	var res strings.Builder
+	vcursor.Session().GetSystemVariables(func(k, v string) {
+		res.WriteString(fmt.Sprintf("SET_VAR(%s = %s) ", k, v))
+	})
+	return strings.TrimSpace(res.String()), nil
```

Comment:
trying to understand if there is a special reason to send as a function to GetSystemVariables

Reply:
This is done to avoid data race while using the system variables map. More details on [this comment thread](https://github.com/vitessio/vitess/pull/9641#discussion_r803656132) :)